### PR TITLE
CSI 4683 servicer for EAR replication

### PR DESCRIPTION
**Pull request by ELENAGER**

### CODE REVIEW

**Review by matancarmeli7 — APPROVED** [review]
LGTM, just create a ticket for messages in the CSI directory

File: controllers/servers/utils.py [hunk 1]
```diff
@@ -606,11 +606,16 @@ def generate_addons_replication_request(request, replication_type):
 
     other_system_id = request.parameters.get(servers_settings.PARAMETERS_SYSTEM_ID)
     copy_type = request.parameters.get(servers_settings.PARAMETERS_COPY_TYPE, REPLICATION_DEFAULT_COPY_TYPE)
+
+    replication_policy = None
+    if servers_settings.PARAMETERS_REPLICATION_POLICY in request.parameters:
+        replication_policy = request.parameters.get(servers_settings.PARAMETERS_REPLICATION_POLICY)
```
Comment (ArbelNathan) [inline controllers/servers/utils.py:612]: if you use `.get` I don't think you need the `if` or the  initialization

File: controllers/array_action/array_mediator_svc.py [hunk 2]
```diff
@@ -280,6 +280,7 @@ def _generate_volume_response(self, cli_volume, is_virt_snap_func=False):
             source_id=source_id,
             array_type=self.array_type,
             space_efficiency_aliases=space_efficiency,
+            volume_group_id=cli_volume.volume_group_id
```
[the remark this answers is not in the document]
Comment (ELENAGER) [reply controllers/array_action/array_mediator_svc.py:283]: The field exists, but it can be empty

File: controllers/servers/utils.py (hunk 1 above)
Comment (ELENAGER) [reply controllers/servers/utils.py:612]: Done

File: controllers/servers/csi/addons_server.py [hunk 3]
```diff
@@ -139,3 +137,22 @@ def ResyncVolume(self, request, context):
...
+        elif replication.replication_type == array_settings.REPLICATION_TYPE_EAR:
+            if replication.volume_group_id != volume.volume_group_id:
+                message = "replication already exists, " \
+                      "but volume {} belongs to another group {}".format(volume.name, volume.volume_group_name)
+                return False, message
+            if replication.name != replication_request.replication_policy:
+                message = "replication already exists, " \
+                          "but volume {} uses another replication policy {}".format(volume.name, replication.name)
+                return False, message
+        return True, ""
```
Comment (ArbelNathan) [inline controllers/servers/csi/addons_server.py:158]: so you can know if the function returns idempotent by the message returns? just `if message:`..

File: controllers/tests/array_action/svc/test_settings.py [hunk 4]
```diff
@@ -121,6 +122,7 @@
 LSRCRELATIONSHIP_AUX_VOLUME_ATTR_KEY = "aux_vdisk_name"
 DUMMY_VOLUME_SITE1 = "volume_site_1"
 DUMMY_VOLUME_SITE2 = "volume_site_2"
+DUMMY_VOLUME_GROUP_ID = "1"
```
Comment (ArbelNathan) [inline controllers/tests/array_action/svc/test_settings.py:125]:
```suggestion
DUMMY_VOLUME_GROUP_ID = "volume_group_id"
```

File: controllers/array_action/array_mediator_svc.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ELENAGER) [reply controllers/array_action/array_mediator_svc.py:283]: Tested on SVC version 8.3.1.7

File: controllers/servers/csi/addons_server.py (hunk 3 above)
Comment (ELENAGER) [reply controllers/servers/csi/addons_server.py:158]: Of course, but it's less elegant from my point of view.

File: controllers/tests/array_action/svc/test_settings.py (hunk 4 above)
Comment (ELENAGER) [reply controllers/tests/array_action/svc/test_settings.py]: Done

File: controllers/array_action/array_mediator_svc.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controllers/array_action/array_mediator_svc.py:283]: how about 7.8.x and 8.2.x?

File: controllers/servers/csi/addons_server.py (hunk 3 above)
Comment (ArbelNathan) [reply controllers/servers/csi/addons_server.py:158]: its a common logic
you can call it "err" instead of "message" if its helps

Comment (ELENAGER) [reply controllers/servers/csi/addons_server.py]: We have no case of True + message, so I just removed flag and checking only message

Comment (ArbelNathan) [reply controllers/servers/csi/addons_server.py]: please name it to error or error_message

Comment (ELENAGER) [reply controllers/servers/csi/addons_server.py]: Done

**Review by ArbelNathan — APPROVED** [review]
LGTM

**Pull request merged**